lib/ukcpio/cpio.c

```diff
@@ -0,0 +1,312 @@
+
+	add_slash = prefix[prefix_len - 1] == '/' ? 0 : 1;
+	abs_path_len = prefix_len + add_slash + path_len + 1;
+
+	abs_path = malloc(abs_path_len);
+
+	if (abs_path == NULL)
+		return NULL;
+
+	memcpy(abs_path, prefix, prefix_len);
+	if (add_slash)
+		abs_path[prefix_len] = '/';
+	memcpy(&abs_path[prefix_len + add_slash], path, path_len);
+
+	abs_path[abs_path_len - 1] = '\0';
+	return abs_path;
+}
+
+/**
+ * Reads the section to the dest from a given a CPIO header.
+ *
+ * @param header_ptr
+ *  Pointer to the CPIO header.
+ * @param dest
+ *  Destination path to extract the current header.
+ * @param last
+ *  The size of the CPIO section.
+ * @return
+ *  Returns 0 on success or one of ukcpio_error enum.
+ */
+static enum ukcpio_error
+read_section(struct cpio_header **header_ptr,
+				const char *dest, uintptr_t last)
+{
+	enum ukcpio_error error = UKCPIO_SUCCESS;
+	int fd;
+	struct cpio_header *header;
+	char *path_from_root;
+	mode_t header_mode;
+	uint32_t header_filesize;
+	uint32_t header_namesize;
+	char *data_location;
+	uint32_t bytes_to_write;
+	int bytes_written;
+	struct cpio_header *next_header;
+
+	if (strcmp(filename(*header_ptr), "TRAILER!!!") == 0) {
+		*header_ptr = NULL;
+		return UKCPIO_SUCCESS;
+	}
+
+	if (!valid_magic(*header_ptr)) {
+		*header_ptr = NULL;
+		return -UKCPIO_INVALID_HEADER;
+	}
+
+	UK_ASSERT(dest);
+
+	header = *header_ptr;
+	path_from_root = absolute_path(dest, filename(header));
+
+	if (path_from_root == NULL) {
+		*header_ptr = NULL;
+		error = -UKCPIO_NOMEM;
+		goto out;
+	}
+
+	header_mode = GET_MODE(header);
+	header_filesize = S8HEX_TO_U32(header->filesize);
+	header_namesize = S8HEX_TO_U32(header->namesize);
+
+	if ((uintptr_t)header + sizeof(struct cpio_header) > last) {
+		*header_ptr = NULL;
+		error = -UKCPIO_MALFORMED_FILE;
+		goto out;
+	}
+
+	if (IS_FILE(header_mode) && header_filesize != 0) {
+		uk_pr_info("Extracting %s...\n", path_from_root);
+		fd = open(path_from_root, O_CREAT | O_RDWR);
+
+		if (fd < 0) {
+			*header_ptr = NULL;
+			error = -UKCPIO_FILE_CREATE_FAILED;
+			goto out;
+		}
+
+		data_location = (char *)ALIGN_4(
+				(char *)(header) + sizeof(struct cpio_header)
+				+ header_namesize);
+
+		if ((uintptr_t)data_location + header_filesize > last) {
+			*header_ptr = NULL;
+			error = -UKCPIO_MALFORMED_FILE;
+			goto out;
+		}
+
+		bytes_to_write = header_filesize;
+		bytes_written = 0;
+
+		while (bytes_to_write > 0) {
+			if ((bytes_written =
+				 write(fd, data_location + bytes_written,
+							 bytes_to_write))
+					< 0) {
+				*header_ptr = NULL;
+				error = -UKCPIO_FILE_WRITE_FAILED;
+				goto out;
+			}
+			bytes_to_write -= bytes_written;
+		}
+
+		if (chmod(path_from_root, header_mode & 0777) < 0)
+			uk_pr_err("Failed to chmod %s\n", path_from_root);
+
+		if (close(fd) < 0) {
+			*header_ptr = NULL;
+			error = -UKCPIO_FILE_CLOSE_FAILED;
+			goto out;
+		}
+	} else if (IS_DIR(header_mode)) {
+		uk_pr_info("Extracting %s...\n", path_from_root);
```

Comment:
Maybe call this message
```c
uk_pr_info("Creating %s...\n", path_from_root);
```
based on the message we are then able to distiguish quicker what happens (extracting a file or creating a directory)